### ISSUE TITLE
Better indicate which utilities are private vs. public

### ISSUE THREAD

**Issue by jrbourbeau** [dask/dask#10119]
We have lots of utilities throughout our codebase and we're not very vigilant about indicating which ones are intended to be public and which are private. Some methods like `dask.utils._deprecated` are marked as private

https://github.com/dask/dask/blob/0cbc46ac89b6f6a2cf949f2fd73c4c1175419db5/dask/utils.py#L78-L85

but most aren't marked with a leading underscore and also aren't meant for public consumption. 

This lack of public vs. private specification causes issues like https://github.com/dask/dask/issues/10112. I think we should move utilities are that are intended to be private into a private module (e.g. `dask._utils`) -- we'll also want to emit deprecation warnings appropriately. 

Note that we do have [some public utilities](https://docs.dask.org/en/stable/api.html#utilities), so this wouldn't be a blanket "move everything into private modules". But I suspect most things in, for example, `dask.utils_test` and `dask.compatibility` can be moved over into private modules (xref https://github.com/dask/dask/pull/10114). 

cc @jacobtomlinson for visibility 

Also cc @fjetter @crusaderky @hendrikmakait as I think the same argument applies to `dask/distributed`

**Comment by jrbourbeau** [comment]
Yeah, agreed that `dask.utils._deprecated` should move to `dask._utils.deprecated` (similarly for other currently private methods). Things that don't have underscores, but are intended to be private should just move into a private module as-is (e.g. `dask.utils.foo` --> `dask._utils.foo`). Public utilities can stay in `dask.utils`. 

> and `distributed`?

I'd welcome thoughts from others here. My current thinking is `dask` and `distributed` are co-developed, so I think `distributed` using private code is fine. If it's not too cumbersome, we could also make utilities `distributed` needs public if we think it's appropriate

**Comment by jacobtomlinson** [comment]
> My current thinking is dask and distributed are co-developed, so I think distributed using private code is fine.

I would agree with this. Although it does provide some CI challenges as changes in one library can break the other.

For a long time, I also considered `dask-{jobqueue,kubernetes,cloudprovider,etc}` co-developed, but we seem to be moving away from that mindset as highlighted by https://github.com/dask/dask/issues/10112. I'm fine with this, but I'd be keen for us to explicitly decide this is the case.

**Comment by crusaderky** [comment]
From memory, we discussed this in the past already.
The consensus was that whatever is documented in sphinx is public API, and whatever isn't is private.
Is there a particular reason to be more aggressive than this in hiding the private API from users?

**Comment by hendrikmakait** [comment]
> The consensus was that whatever is documented in sphinx is public API, and whatever isn't is private.

I don't know how intuitive this is to people. Until this rule was explained to me, I did not realize this was how Dask differentiates between public and private API. I had assumed that we were simply not very diligent in differentiating between the two, and documentation varied based on the usage frequency. Additionally, there are parts of the private API that I would very much like to be documented in detail to understand the underlying implementation, this would break the rule.
